pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java

```diff
@@ -2220,20 +2219,19 @@ private CompletableFuture<Integer> unloadServiceUnit(NamespaceBundle serviceUnit
             if (serviceUnit.includes(topicName)) {
                 // Topic needs to be unloaded
                 log.info("[{}] Unloading topic", topicName);
-                if (topicFuture.isCompletedExceptionally()) {
-                    try {
-                        topicFuture.get();
-                    } catch (InterruptedException | ExecutionException ex) {
-                        if (ex.getCause() instanceof ServiceUnitNotReadyException) {
-                            // Topic was already unloaded
-                            if (log.isDebugEnabled()) {
-                                log.debug("[{}] Topic was already unloaded", topicName);
-                            }
-                            return;
-                        } else {
-                            log.warn("[{}] Got exception when closing topic", topicName, ex);
-                        }
+                if (!topicFuture.isDone()) {
+                    // Fail-fast unfinished topic creation.
+                    if(topicFuture.completeExceptionally(
+                            new ServiceUnitNotReadyException("Topic creation has been canceled by not ready."))
+                            && log.isDebugEnabled()) {
+                        log.debug("[{}][{}] Cancel the topic creation when unloading the service unit.",
+                                serviceUnit, name);
                     }
+                    return;
+                }
+                // Ignore exceptional topic
+                if (topicFuture.isCompletedExceptionally()) {
+                    return;
```

Comment:
But we might miss the topic close if the topic creation is almost done, right?
Should we also check the topicFuture when creating the topic? If the topicFuture is already completed, we should close that topic.

Reply:
Improved it.